src/test/fuzz/versionbits.cpp

```diff
@@ -0,0 +1,345 @@
+    int64_t BeginTime(const Consensus::Params& params) const override { return m_begin; }
+    int64_t EndTime(const Consensus::Params& params) const override { return m_end; }
+    int Period(const Consensus::Params& params) const override { return m_period; }
+    int Threshold(const Consensus::Params& params) const override { return m_threshold; }
+
+    ThresholdState GetStateFor(const CBlockIndex* pindexPrev) const { return AbstractThresholdConditionChecker::GetStateFor(pindexPrev, dummy_params, m_cache); }
+    int GetStateSinceHeightFor(const CBlockIndex* pindexPrev) const { return AbstractThresholdConditionChecker::GetStateSinceHeightFor(pindexPrev, dummy_params, m_cache); }
+    BIP9Stats GetStateStatisticsFor(const CBlockIndex* pindexPrev) const { return AbstractThresholdConditionChecker::GetStateStatisticsFor(pindexPrev, dummy_params); }
+
+    bool Condition(int64_t version) const
+    {
+        return ((version >> m_bit) & 1) != 0 && (version & VERSIONBITS_TOP_MASK) == VERSIONBITS_TOP_BITS;
+    }
+
+    bool Condition(const CBlockIndex* pindex) const { return Condition(pindex->nVersion); }
+};
+
+/** Track blocks mined for test */
+class Blocks
+{
+private:
+    std::vector<std::unique_ptr<CBlockIndex>> m_blocks;
+    const uint32_t m_start_time;
+    const uint32_t m_interval;
+    const int32_t m_signal;
+    const int32_t m_no_signal;
+
+public:
+    Blocks(uint32_t start_time, uint32_t interval, int32_t signal, int32_t no_signal)
+        : m_start_time{start_time}, m_interval{interval}, m_signal{signal}, m_no_signal{no_signal} {}
+
+    size_t size() const { return m_blocks.size(); }
+
+    CBlockIndex* tip() const
+    {
+        return m_blocks.empty() ? nullptr : m_blocks.back().get();
+    }
+
+    CBlockIndex* mine_block(bool signal)
+    {
+        CBlockHeader header;
+        header.nVersion = signal ? m_signal : m_no_signal;
+        header.nTime = m_start_time + m_blocks.size() * m_interval;
+        header.nBits = 0x1d00ffff;
+
+        auto current_block = std::make_unique<CBlockIndex>(header);
+        current_block->pprev = tip();
+        current_block->nHeight = m_blocks.size();
+        current_block->BuildSkip();
+
+        return m_blocks.emplace_back(std::move(current_block)).get();
+    }
+};
+
+void initialize()
+{
+    SelectParams(CBaseChainParams::MAIN);
+}
+} // namespace
+
+constexpr uint32_t MAX_TIME = 4102444800; // 2100-01-01
+
+FUZZ_TARGET_INIT(versionbits, initialize)
+{
+    const CChainParams& params = Params();
+
+    const int64_t interval = params.GetConsensus().nPowTargetSpacing;
+    assert(interval > 1); // need to be able to halve it
+    assert(interval < std::numeric_limits<int32_t>::max());
+
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+
+    // making period/max_periods larger slows these tests down significantly
+    const int period = 32;
+    const size_t max_periods = 16;
+    const size_t max_blocks = 2 * period * max_periods;
```

Comment:
I don't understand the factor of 2 here (or lower down in `if (blocks.size() + period*2 > max_blocks)`. What is it needed for?

Reply:
`max_blocks` is the max number of blocks "mined"; there shouldn't be any point mining more than `period*max_period` worth -- things should just be stuck in ACTIVE/FAILED from that point on, but this lets some extra amount be checked just in case.

If `blocks.size() + period*2 > max_blocks` then an additional round of the loop and then the final period combined would end up mining more than `max_blocks` in total.

Reply:
Ah, got it. So `max_periods` is the max number of periods in the versionbits parameters and `max_blocks` is the max number of blocks in the test.